File: backend/controller/vote.go
package controller

import (
	"net/http"
	"strconv"
	"vote-system-backend/dto"
	"vote-system-backend/service"
	"vote-system-backend/utils"

	"github.com/gin-gonic/gin"
)

type VoteController struct {
	voteService *service.VoteService
}

func NewVoteController() *VoteController {
	return &VoteController{
		voteService: service.NewVoteService(),
	}
}

func (ctrl *VoteController) CreateVote(c *gin.Context) {
	var req dto.CreateVoteRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.Error(c, http.StatusBadRequest, "参数错误")
		return
	}

	userID, exists := c.Get("user_id")
	if !exists {
		utils.Error(c, http.StatusUnauthorized, "未授权")
		return
	}

	vote, err := ctrl.voteService.CreateVote(&req, userID.(uint))
	if err != nil {
		utils.Error(c, http.StatusInternalServerError, err.Error())
		return
	}

	utils.SuccessWithMessage(c, "投票创建成功", gin.H{"id": vote.ID})
}

func (ctrl *VoteController) GetVote(c *gin.Context) {
	idStr := c.Param("id")
	id, err := strconv.ParseUint(idStr, 10, 32)
	userID := c.GetUint("user_id")
	if err != nil {
		utils.Error(c, http.StatusBadRequest, "无效的ID")
		return
	}

	vote, err := ctrl.voteService.GetVote(uint(id), userID)
	if err != nil {
		utils.Error(c, http.StatusBadRequest, err.Error())
		return
	}

	utils.Success(c, vote)
}

func (ctrl *VoteController) UpdateVote(c *gin.Context) {
	var req dto.UpdateVoteRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.Error(c, http.StatusBadRequest, "参数错误")
		return
	}

	userID, exists := c.Get("user_id")
	if !exists {
		utils.Error(c, http.StatusUnauthorized, "未授权")
		return
	}

	if err := ctrl.voteService.UpdateVote(&req, userID.(uint)); err != nil {
		utils.Error(c, http.StatusBadRequest, err.Error())
		return
	}

	utils.SuccessWithMessage(c, "投票更新成功", nil)
}

func (ctrl *VoteController) DeleteVote(c *gin.Context) {
	idStr := c.Param("id")
	id, err := strconv.ParseUint(idStr, 10, 32)
	if err != nil {
		utils.Error(c, http.StatusBadRequest, "无效的ID")
		return
	}

	userID, exists := c.Get("user_id")
	if !exists {
		utils.Error(c, http.StatusUnauthorized, "未授权")
		return
	}

	if err := ctrl.voteService.DeleteVote(uint(id), userID.(uint)); err != nil {
		utils.Error(c, http.StatusBadRequest, err.Error())
		return
	}

	utils.SuccessWithMessage(c, "删除成功", nil)
}

func (ctrl *VoteController) Vote(c *gin.Context) {
	var req dto.VoteRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.Error(c, http.StatusBadRequest, "参数错误")
		return
	}

	userID, exists := c.Get("user_id")
	if !exists {
		utils.Error(c, http.StatusUnauthorized, "未授权")
		return
	}

	if err := ctrl.voteService.Vote(&req, userID.(uint)); err != nil {
		utils.Error(c, http.StatusBadRequest, err.Error())
		return
	}

	utils.SuccessWithMessage(c, "投票成功", nil)
}

func (ctrl *VoteController) GetUserVotes(c *gin.Context) {
	userID, exists := c.Get("user_id")
	if !exists {
		utils.Error(c, http.StatusUnauthorized, "未授权")
		return
	}

	votes, err := ctrl.voteService.GetUserVotes(userID.(uint))
	if err != nil {
		utils.Error(c, http.StatusInternalServerError, err.Error())
		return
	}

	utils.Success(c, votes)
}

func (ctrl *VoteController) GetAllVotes(c *gin.Context) {
	votes, err := ctrl.voteService.GetAllVotes()
	if err != nil {
		utils.Error(c, http.StatusInternalServerError, err.Error())
		return
	}

	utils.Success(c, votes)
}

File: frontend/src/services/vote.ts
import type { Vote, CreateVoteRequest, UpdateVoteRequest, VoteRequest } from '@/types/vote';
import type { Response } from '@/types/api';
import api from './api';

const BASE_URL = '/vote';

/**
 * 创建投票
 * @param data 创建投票数据
 * @returns Promise<Vote>
 */
export const createVote = async (data: CreateVoteRequest): Promise<Response<Vote>> =>
  api.post(`${BASE_URL}/create`, data);

/**
 * 获取投票详情
 * @param id 投票ID
 * @returns Promise<VoteResponse>
 */
export const getVote = async (id: string | number): Promise<Response<Vote>> => api.get(`${BASE_URL}/${id}`);

/**
 * 更新投票
 * @param data 更新投票数据
 * @returns Promise<Vote>
 */
export const updateVote = async (data: UpdateVoteRequest): Promise<Vote> => api.put(`${BASE_URL}/update`, data);

/**
 * 删除投票
 * @param id 投票ID
 * @returns Promise<void>
 */
export const deleteVote = (id: string | number) => api.delete(`${BASE_URL}/${id}`);

/**
 * 投票
 * @param data 投票数据
 */
export const submitVote = (data: VoteRequest): Promise<Response<void>> => api.post(`${BASE_URL}/submit`, data);

/**
 * 获取用户的投票列表
 */
export const getUserVotes = () => api.get(`${BASE_URL}/my`);

/**
 * 获取所有投票列表
 */
export const getAllVotes = (): Promise<Response<Vote[]>> => api.get(`${BASE_URL}/all`);
